pulp_container/app/tasks/sync_stages.py

```diff
@@ -112,105 +153,110 @@ async def run(self):
-                dl_res = await download_tag
-                with open(dl_res.path, "rb") as content_file:
-                    raw_data = content_file.read()
-                dl_res.artifact_attributes["file"] = dl_res.path
-                saved_artifact = await sync_to_async(_save_artifact_blocking)(
-                    dl_res.artifact_attributes
-                )
+            async with ProgressReport(
+                message="Downloading manifests", code="sync.downloading.manifests"
+            ) as manifest_pr:
 
-                tag_name = dl_res.url.split("/")[-1]
-                tag_dc = DeclarativeContent(Tag(name=tag_name))
+                to_download_artifact = [
+                    self._check_for_existing_manifest(download_tag, progress_report=manifest_pr)
+                    for download_tag in asyncio.as_completed(to_download)
+                ]
 
-                content_data = json.loads(raw_data)
-                media_type = determine_media_type(content_data, dl_res)
-                digest = dl_res.artifact_attributes["sha256"]
-                validate_manifest(content_data, media_type, digest)
+                for artifact in asyncio.as_completed(to_download_artifact):
+                    saved_artifact, content_data, raw_data, response = await artifact
 
-                if media_type in (MEDIA_TYPE.MANIFEST_LIST, MEDIA_TYPE.INDEX_OCI):
-                    list_dc = self.create_tagged_manifest_list(
-                        tag_name, saved_artifact, content_data, media_type
-                    )
-                    for listed_manifest_task in asyncio.as_completed(
-                        [
-                            self.create_listed_manifest(manifest_data)
-                            for manifest_data in content_data.get("manifests")
-                        ]
-                    ):
-                        listed_manifest = await listed_manifest_task
-                        man_dc = listed_manifest["manifest_dc"]
-                        if signature_source is not None:
-                            man_sig_dcs = await self.create_signatures(man_dc, signature_source)
-                            if self.signed_only and not man_sig_dcs:
-                                log.info(
-                                    "The unsigned image {img_digest} which is a part of the "
-                                    "manifest list {ml_digest} (tagged as `{tag}`) can't be "
-                                    "synced due to a requirement to sync signed content only. "
-                                    "The whole manifest list is skipped.".format(
-                                        img_digest=man_dc.content.digest,
-                                        ml_digest=list_dc.content.digest,
-                                        tag=tag_name,
+                    media_type = determine_media_type(content_data, response)
+                    validate_manifest(content_data, media_type, response)
```

Comment:
The `validate_manifest` function expects the digest of a manifest instead of `response`. It is used for logging purposes: https://github.com/pulp/pulp_container/blob/0737a90cea8978777030bc7a8550f93b3d29827b/pulp_container/app/utils.py#L198-L212

Reply:
Thanks man. Don't know why but the `test_sync` suite didn't fail with it.

Reply:
That makes sense if there were no errors raised. So, it did not reach the branch which uses `digest`.

This is the place where the `typing` module would be useful. 🌺

Reply:
@lubosmj  speaking of https://github.com/pulp/pulp_container/issues/1130